### PR TITLE
Medical Equipment section Parsing

### PR DESCRIPTION
**Pull request by kushalvermaecw**
This PR includes CCDA parsing for Medical Equipment section. This will include: 
1. Medical Equipment Section parsing is added which parses medical equipment supply organizer, implants during procedure and non medical supply activity details.
2. Test cases are added to validate parsing.
3. Added spring-test framework dependency in` pom.xml` as `ProcedureProcessor` needed to be autowired in `MedicalEquipmentProcessor `.
4. Updated `pom.xml` to update jUnit version from 4.11 to 4.12 to make jUnit compatible with spring-test framework.
5. Added sample CCDA document for Medical Equipment section for fictitious patient.

### CODE REVIEW

**Comment by kushalvermaecw** [comment]
@drbgfc ,

We have created Pull request for Medical Equipment Section. Please review it.

**Comment by drbgfc** [comment]
Hi  Apologies - I have been OOO. I will look at this soon. Thanks!

File: src/main/java/org/sitenv/ccdaparsing/model/CCDAMedicalEquipment.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    public List<CCDAProcActProc> getProcedureActs() {
+        return procedureActs == null ? new ArrayList<CCDAProcActProc>() : new ArrayList<>(procedureActs);
+    }
+    public ArrayList<String> getReferenceLinks() {
+        return referenceLinks == null ? new ArrayList<String>() : new ArrayList<>(referenceLinks);
+    }
+    public void setReferenceLinks(ArrayList<String> referenceLinks) {
+        this.referenceLinks = referenceLinks == null ? new ArrayList<String>() : new ArrayList<>(referenceLinks);
+    }
+}
```
Comment (drbgfc) [inline src/main/java/org/sitenv/ccdaparsing/model/CCDAMedicalEquipment.java:90]: Any particular reason we don't override equals and Code for this class?

File: src/main/java/org/sitenv/ccdaparsing/model/CCDAMedicalEquipmentOrg.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
+package org.sitenv.ccdaparsing.model;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
```
Comment (drbgfc) [inline src/main/java/org/sitenv/ccdaparsing/model/CCDAMedicalEquipmentOrg.java:6]: You might want your comment here like the other classes to identify it was your work?

File: src/main/java/org/sitenv/ccdaparsing/model/CCDARefModel.java [hunk 3]
```diff
@@ -154,6 +155,12 @@ public CCDAFamilyHistory getFamilyHistory() {
 		return familyHistory;
 	}
 
+	public void setMedicationEquipments(CCDAMedicalEquipment medicalEquipment) {
```
Comment (drbgfc) [inline src/main/java/org/sitenv/ccdaparsing/model/CCDARefModel.java:158]: do you mind if we reweite this as:

```java
public void setMedicalEquipment(CCDAMedicalEquipment medicalEquipment) {**
	this.medicalEquipment = medicalEquipment;
}
```

Equipments is technically not a word. Nor does its Medication prefix match the variable name, medicalEquipment. Can we also rename the getter in the same manner to:
```java
public CCDAMedicalEquipment getMedicalEquipment() {
```
If this does not make sense for some reason (using MedicalEquipment). Can we then rename the class itself, to CCDAMedicationEquipment vs CCDAMedicalEquipment and then remove the plural form on the getter and the setter.
Let me know if any questions or disagreements. Thanks

File: src/main/java/org/sitenv/ccdaparsing/processing/MedicalEquipmentProcessor.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+@Service
+public class MedicalEquipmentProcessor {
+    private static final Logger logger = Logger.getLogger(MedicalEquipmentProcessor.class);
+
+    @Autowired
+    ProcedureProcessor procedureProcessor;
+
+    @Async()
+    public Future<CCDAMedicalEquipment> retrieveMedicalEquipments(XPath xPath , Document doc) throws XPathExpressionException, TransformerException {
```
Comment (drbgfc) [inline src/main/java/org/sitenv/ccdaparsing/processing/MedicalEquipmentProcessor.java:35]: A few more examples of 'equipments' in this class which I think should be singular. I don't want to be pedantic, so, if you don't agree, don't worry about updating.

File: src/main/java/org/sitenv/ccdaparsing/util/ApplicationConstants.java [hunk 5]
```diff
@@ -76,6 +76,12 @@
 	
 	public final static String LINE_NUMBER_KEY_NAME = "lineNumber";
     public final static String END_LINE_NUMBER_KEY_NAME = "endLineNumber";
+
```
Comment (drbgfc) [inline src/main/java/org/sitenv/ccdaparsing/util/ApplicationConstants.java:79]: I like this addition.

**Review by drbgfc — CHANGES_REQUESTED** [review]
Great work, thanks.
I added some inline comments. Most are just that, comments. The only one I'd request to fix is the setMedicationEquipments and getMedicationEquipments type as it actually should be set and get MedicalEquipment. Thanks again.

File: src/main/java/org/sitenv/ccdaparsing/util/ApplicationConstants.java [hunk 6]
```diff
@@ -76,6 +76,12 @@
...
+	public static final String TEMPLATE_ID_EXPRESSION      = "./templateId[not(@nullFlavor)]";
+	public static final String CODE_EXPRESSION             = "./code[not(@nullFlavor)]";
+	public static final String EFFECTIVE_EXPRESSION        = "./effectiveTime[not(@nullFlavor)]";
+	public static final String ID_EXPRESSION               = "./id[not(@nullFlavor)]";
+	public static final String STATUS_EXPRESSION           = "./statusCode[not(@nullFlavor)]";
```
Comment (drbgfc) [inline src/main/java/org/sitenv/ccdaparsing/util/ApplicationConstants.java:84]: I like this addition.

**Comment by drbgfc** [comment]
If you're intent was to just share this code, then I will update as needed, not trying to take your time. If your intent was to merge so that we could have the same codebase, let me know.

**Comment by kushalvermaecw** [comment]
> If you're intent was to just share this code, then I will update as needed, not trying to take your time. If your intent was to merge so that we could have the same codebase, let me know.

Hello Dan,
We will be making changes suggested and let you know

File: src/main/java/org/sitenv/ccdaparsing/model/CCDAMedicalEquipment.java (hunk 1 above)
Comment (kushalvermaecw) [reply src/main/java/org/sitenv/ccdaparsing/model/CCDAMedicalEquipment.java:90]: This was missed out we will add it

File: src/main/java/org/sitenv/ccdaparsing/model/CCDARefModel.java (hunk 3 above)
Comment (kushalvermaecw) [reply src/main/java/org/sitenv/ccdaparsing/model/CCDARefModel.java]: we have made those changes

**Comment by kushalvermaecw** [comment]
> > If you're intent was to just share this code, then I will update as needed, not trying to take your time. If your intent was to merge so that we could have the same codebase, let me know.
> 
> Hello Dan,
> We will be making changes suggested and let you know

Hello Dan,
We have completed the changes you recommended. Please review them

**Review by drbgfc — APPROVED** [review]

**Comment by drbgfc** [comment]
> > > If you're intent was to just share this code, then I will update as needed, not trying to take your time. If your intent was to merge so that we could have the same codebase, let me know.
> > 
> > 
> > Hello Dan,
> > We will be making changes suggested and let you know
> 
> Hello Dan,
> We have completed the changes you recommended. Please review them

Thanks!

**Pull request merged**